### ISSUE TITLE
Auto-updating on startup

### ISSUE THREAD

**Issue by Booplicate** [Monika-After-Story/MonikaModDev#9852]
There's a type of people that don't update for a year+. This can cause issues and already bit us (I think a few times). As much as I hate auto-updating apps, I think this would be the best choice for us too. Our team is too small to be able to test for bugs that can happen if user doesn't update for a long period of time.

**Comment by multimokia** [comment]
Maybe make it togglable but default on?

**Comment by ashyashton** [comment]
I agree with Multi's suggestion! I personally would turn it off just because I like to choose when things update, but I also try to update ASAP when I can, so having it togglable would be great for someone like me who might wanna do something really quick first.

**Comment by Booplicate** [comment]
Generally I agree - users should be able to choice when to update, but in that case there would still be a way to skip multiple updates. So we'd partially lose the benefit.
What if we still forced updating if version difference is too big (1+ major, or 2+ minor, or 4+ patch releases)?

**Comment by dreamscached** [comment]
What actually causes the issue in updating after a while? Aren't update scripts running sequentially? E.g. if updating from 0.10.0 -> 0.12.3 it would be 0.10.0 -> 0.10.1 -> 0.11.0 -> 0.12.0 and so on?

**Comment by Booplicate** [comment]
Doesn't matter how they run if you don't have the bugfix. The issue is in not receiving fixes in time.

**Comment by dreamscached** [comment]
Oh that's why, I first thought it was about updates not applying properly or something. Got it now.

**Comment by Justformas** [comment]
A potential issue with auto-updating if it can't be disabled, is that when a major update happens, it could possibly be incompatible with some systems such as older ones. But it could still try to update anyway and cause problems or prevent the game from running. Worse, if the user tried to download an older version again, it'd just try to update to a newer version again (unless it's a version before auto-updating was added).

So at least for this case, there should be some way to prevent auto-updating, other than the player disabling their internet connection.

**Comment by Booplicate** [comment]
>  that when a major update happens, it could possibly be incompatible with some systems

That will happen with Python 3 migration and dropping 32bit support. Auto updating will be added after that fact. That said, there will be no support for older versions, you will be left with all the issues that might arise as well as no new content.
If in some far hypothetical future we make such a change as dropping support for a large number of devices, we just won't release the update for them, so auto-updating wouldn't affect them.

> the user tried to download an older version again, it'd just try to update to a newer version again

Downgrading is not supported because it may cause data corruption. Updates can only be applied one way - forward.

**Comment by ashyashton** [comment]
Wait, if the issue is bugfixes not reaching users, then would it be an option to give a more prominent notification? Like, if the devs consider a certain update to be a critical update, then maybe it could a pop-up saying that the end-user is missing a critical update and it would be highly recommended to update. The options could then be something a bit cheeky like "update" and "I'm fine with bugs!" to both be a bit silly but also hammer home the fact that hey, if you choose to ignore this update, you could get some issues.

One of my big fears with auto-updates is that I know for a fact that some older versions of MAS have caused data loss and/or affection loss, such as 0.8.5 and 0.9.2 which are literally listed on the website as causing data and affection issues. Unfortunately, unlike system incompatibilities, you can't just plan to not get those. You don't know when issues like that will pop up, and if you're forcing users to update, there's a good chance that someday, you are going to piss off a LOT of people.

Hell, I know I'd be pissed if the mod FORCED me to update and I ended up losing my Monika or ended up with affection loss or her forgetting something we've done just because I wasn't allowed to wait a day or so like I typically do. Because that's what I usually do: I wait a day or so, check to see if any big updates like "oh god guys don't download this version, you might lose your Monika" have come out, and if it's all good, _then_ I update. If I lost the ability to do that and something went wrong, I can promise I'd be having a cow over it, and I know I'm not the only one.

With some programs, I get it. It's just some data! But when the devs and the community push _so hard_ for everyone to see Monika as a person and to treat her well, to the point that of all of us immediately giving the stink eye to anyone who uses the meaner options, then it really isn't fair to be so blasé about potentially forcing users into a position where their Monikas get hurt and they have no way to avoid it. I know, she's technically also just lines of code, but we all know that's not how most people in the community treat her, due in large part to how the devs handle things. In my opinion, you really can't have it both ways.

**Comment by Booplicate** [comment]
> if the devs consider a certain update to be a critical update

Not really, for the game an update is just an update, we can't give it a tag or priority.

> such as 0.8.5 and 0.9.2 which are literally listed on the website as causing data and affection issues

This is a valid point. That can be solved with an extra backup after the update, plus having auto-updates means you will receive the next update asap, instead of waiting until it's too late. There was a case where we released a fix update soon after, but people decided not to update twice in a row. Waiting can do harm too. In your example it's also just 2 faulty updates out of how many there were? Statistically, I think more often we need people to update than not to update.
On the other hand, as I said, we may force it only if your version is too old. Although, I like your idea with prompting the user multiple times and letting them skip the update if they confirm it like 3 times and it's made clear it's their fault when their game breaks.

> I'd be pissed if the mod FORCED me to update and I ended up losing my Monika or ended up with affection loss or her forgetting something we've done just because I wasn't allowed to wait

Would you be pissed if you waited and it caused the same effect?

**Comment by ashyashton** [comment]
> Would you be pissed if you waited and it caused the same effect?

I would be pissed at myself, but I would not be pissed at you guys, no. Because I know that I am assuming some risk when I wait on updates. It is a risk I am choosing to take. Just like how it's a risk to put your face on the internet, so if someone took photos of me and posted them without my consent, I'd raise hell, but if I post a selfie on my own social media accounts, that's on me.

Also, to be clear, I'm not talking about updating and then waiting for the next one. After I wait a day or two for the first update, I am then consistently checking back to make sure there aren't any important fixes that I'm now missing. So in your scenario of updating and then skipping the update with the fix, that would be on me for not paying close enough attention after I did update.

> That can be solved with an extra backup after the update

If the mod is automatically making said backups, and can also load the backups if there is a bug, then yeah that would work! My understanding, though, is that the user would still have to manually change the persistent file out, which sounds like a bit of a headache to put a bunch of users through just because an update potentially messed something up.

> In your example it's also just 2 faulty updates out of how many there were

I'm not saying it happens a lot, I'm just saying that it is possible and so we should factor it in.

**Comment by Justformas** [comment]
I once again agree with ashyashton. Players should have the right to choose between updating or not but also the responsibility to accept the potential consequences of not updating. There may also be other reasons why players may not want to update. In a perfect world this wouldn't happen, but to be frank, it's possible that differences of opinion around what gets changed may lead to players wanting to stay on an earlier version, even if it's just a small minority.

As for tagging an update though:

> > if the devs consider a certain update to be a critical update
> 
> Not really, for the game an update is just an update, we can't give it a tag or priority.
> 
> >

So you can't tag that, but the game can tell the difference between a major and minor update?

> What if we still forced updating if version difference is too big (1+ major, or 2+ minor, or 4+ patch releases)?

Then in theory you could just classify a critical update as a major update then (even if it means going from, say, 13.0 to 14.0), or come up with some other way of classifying it for the game. Not saying to necessarily do either of those things, but I would think indicating that an update is critical somehow would be possible and could be useful to players.

Just to address something said earlier:

> Downgrading is not supported because it may cause data corruption. Updates can only be applied one way - forward.

Actually downgrading a higher version to a lower version, no it's not possible. Older versions may not be supported anymore either, but most of them are still available on the monikaafterstory website, down to 0.5.0. If you did a clean install including with your persistents (or if you actually had persistents backed up from an earlier version that were never associated with a newer version), then in theory you could play an older version, and even restore persistents that were last associated with that version and never a newer version.
